### PR TITLE
client-go: introduce feature gates

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What this PR does / why we need it:
TBD

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
Heavily based on https://github.com/kubernetes/kubernetes/pull/120972
All credit goes to @deads2k 
#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/features/kube_features.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+// DefaultFeatureGates returns the feature gates exposed by this library.
+//
+// By default, only the default features gate will be returned.
+// The default implementation allows controlling the features
+// via environmental variables.
+// For example, if you have a feature named "MyFeature,"
+// setting an environmental variable "KUBE_FEATURE_MyFeature"
+// will allow you to configure the state of that feature.
+//
+// Please note that the actual set of the features gates
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/client-go/features/kube_features.go:32]: note for myself: mention that implementation can be overwritten as well.

File: staging/src/k8s.io/client-go/features/kube_features.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	return mutableFeatureGates.Add(defaultKubernetesFeatureGates)
+}
+
+// SetFeatureGates overwrites the default implementation of the feature gate
+// used by this library.
+//
+// Useful for binaries that would like to have full control of the features
+// exposed by this library. For example to allow consumers of a binary
+// to interact with the features via a command line flag.
+func SetFeatureGates(newFeatureGates featuregate.FeatureGate) {
```
Comment (benluddy) [inline staging/src/k8s.io/client-go/features/kube_features.go:51]: Is it always wrong to pass this a FeatureGate without first registering all the features defined in this package? The panics from calls to Enabled could be sneaky. Maybe the call to SetFeatureGates could detect this condition and error, or something similar?

File: staging/src/k8s.io/client-go/features/kube_features.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+// By default, only the default features gate will be returned.
+// The default implementation allows controlling the features
+// via environmental variables.
+// For example, if you have a feature named "MyFeature,"
+// setting an environmental variable "KUBE_FEATURE_MyFeature"
+// will allow you to configure the state of that feature.
+//
+// Please note that the actual set of the features gates
+// might be overwritten by calling SetFeatureGates method.
+func DefaultFeatureGates() featuregate.Reader {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/features/kube_features.go:34]: It depends on the functionality we want to have in the default implementation. If the default implementation is meant to only allow setting flags from environment variables, then you are right that it won't work.

On the other hand, if we are adding a feature gate to the `client-go` library, why can't we add a new test method that allows setting flags in the library?

A new testing method could replace the default implementation with one that allows overwriting/setting flags regardless of environment variables being read or not.

**Comment by p0lyn0mial** [comment]
I talked to @benluddy , and we want to try one more approach, so let's not review this code for now.

**Comment by p0lyn0mial** [comment]
/close

I received an out-of-band feedback on the feature gate mechanism from @deads2k.
We would like to continue with https://github.com/kubernetes/kubernetes/pull/121012

**Pull request closed without merging**